File: src/main/scala/resonantinduction/mechanical/process/crusher/TileMechanicalPiston.java
package resonantinduction.mechanical.process.crusher;

import cpw.mods.fml.common.Loader;
import cpw.mods.fml.relauncher.ReflectionHelper;
import resonant.content.factory.resources.RecipeType;
import resonantinduction.mechanical.energy.grid.MechanicalNode;
import resonantinduction.mechanical.energy.grid.TileMechanical;
import net.minecraft.block.Block;
import net.minecraft.block.material.Material;
import net.minecraft.item.ItemStack;
import net.minecraft.nbt.NBTTagCompound;
import net.minecraft.tileentity.TileEntity;
import net.minecraft.world.World;
import net.minecraftforge.common.util.ForgeDirection;
import resonant.api.IRotatable;
import resonant.api.recipe.MachineRecipes;
import resonant.api.recipe.RecipeResource;
import resonant.lib.config.Config;
import resonant.lib.utility.MovementUtility;
import resonant.lib.utility.inventory.InventoryUtility;
import resonantinduction.core.ResonantInduction;
import universalelectricity.api.core.grid.INode;
import universalelectricity.core.transform.vector.Vector3;

import java.lang.reflect.Method;
import java.util.List;

/**
 * Mechanical driven piston that can be used to move basic blocks and crush ores
 * @author Calclavia
 */
public class TileMechanicalPiston extends TileMechanical
{
    @Config
    private static int mechanicalPistonMultiplier = 2;

    protected boolean markRevolve = false;

    public TileMechanicalPiston()
    {
        super(Material.piston);
        mechanicalNode = new NodeMechanicalPiston(this);
        isOpaqueCube(false);
        normalRender(false);
        customItemRender(true);
        rotationMask_$eq(Byte.parseByte("111111", 2));
        setTextureName("material_steel_dark");
    }

    @Override
    public void getNodes(List<INode> nodes)
    {
        if(mechanicalNode != null)
            nodes.add(this.mechanicalNode);
    }

    @Override
    public void update()
    {
        super.update();

        if (markRevolve)
        {
            Vector3 movePosition = new Vector3(TileMechanicalPiston.this).add(getDirection());

            if (!hitOreBlock(movePosition))
            {
                if (!worldObj.isRemote)
                {
                    Vector3 moveNewPosition = movePosition.clone().add(getDirection());

                    if (canMove(movePosition, moveNewPosition))
                    {
                        move(movePosition, moveNewPosition);
                    }
                }
            }

            markRevolve = false;
        }
    }

    public boolean hitOreBlock(Vector3 blockPos)
    {
        //TODO add a crushing head to enforce block breaking on all block types
        Block block = blockPos.getBlock(world());

        if (block != null)
        {
            ItemStack blockStack = new ItemStack(block);
            RecipeResource[] resources = MachineRecipes.INSTANCE.getOutput(RecipeType.CRUSHER.name(), blockStack);

            if (resources.length > 0)
            {
                if (!worldObj.isRemote)
                {
                    for (RecipeResource recipe : resources)
                    {
                        if (Math.random() <= recipe.getChance())
                        {
                            InventoryUtility.dropItemStack(world(), blockPos.clone().add(0.5), recipe.getItemStack(), 10, 0);
                        }
                    }

                    blockPos.setBlockToAir(world());
                }

                ResonantInduction.proxy().renderBlockParticle(worldObj, blockPos.clone().add(0.5), new Vector3((Math.random() - 0.5f) * 3, (Math.random() - 0.5f) * 3, (Math.random() - 0.5f) * 3), Block.getIdFromBlock(block), 1);
                return true;
            }
        }

        if (!worldObj.isRemote)
        {
            world().destroyBlockInWorldPartially(0, blockPos.xi(), blockPos.yi(), blockPos.zi(), -1);
        }

        return false;
    }

    public boolean canMove(Vector3 from, Vector3 to)
    {
        if (this.equals(to.getTileEntity(getWorldObj())))
        {
            return false;
        }

        /** Check Target */
        Block targetBlock = to.getBlock(worldObj);

        if (!(worldObj.isAirBlock(to.xi(), to.yi(), to.zi()) || (targetBlock != null && (targetBlock.canBeReplacedByLeaves(worldObj, to.xi(), to.yi(), to.zi())))))
        {
            return false;
        }

        return true;
    }

    public void move(Vector3 from, Vector3 to)
    {
        Block blockID = from.getBlock(worldObj);
        int blockMetadata = from.getBlockMetadata(worldObj);

        TileEntity tileEntity = from.getTileEntity(worldObj);

        NBTTagCompound tileData = new NBTTagCompound();

        if (tileEntity != null)
        {
            tileEntity.writeToNBT(tileData);
        }

        MovementUtility.setBlockSneaky(worldObj, from, null, 0, null);

        if (tileEntity != null && tileData != null)
        {
            /** Forge Multipart Support. Use FMP's custom TE creator. */
            boolean isMultipart = tileData.getString("id").equals("savedMultipart");

            TileEntity newTile = null;

            if (isMultipart)
            {
                try
                {
                    Class multipart = Class.forName("codechicken.multipart.MultipartHelper");
                    Method m = multipart.getMethod("createTileFromNBT", World.class, NBTTagCompound.class);
                    newTile = (TileEntity) m.invoke(null, worldObj, tileData);
                }
                catch (Exception e)
                {
                    e.printStackTrace();
                }
            }
            else
            {
                newTile = TileEntity.createAndLoadEntity(tileData);
            }

            MovementUtility.setBlockSneaky(worldObj, to, blockID, blockMetadata, newTile);

            if (newTile != null && isMultipart)
            {
                try
                {
                    // Send the description packet of the TE after moving it.
                    Class multipart = Class.forName("codechicken.multipart.MultipartHelper");
                    multipart.getMethod("sendDescPacket", World.class, TileEntity.class).invoke(null, worldObj, newTile);

                    // Call onMoved event.
                    Class tileMultipart = Class.forName("codechicken.multipart.TileMultipart");
                    tileMultipart.getMethod("onMoved").invoke(newTile);
                }
                catch (Exception e)
                {
                    e.printStackTrace();
                }
            }
        }
        else
        {
            MovementUtility.setBlockSneaky(worldObj, to, blockID, blockMetadata, null);
        }

        notifyChanges(from);
        notifyChanges(to);
    }

    public void notifyChanges(Vector3 pos)
    {
        worldObj.notifyBlocksOfNeighborChange(pos.xi(), pos.yi(), pos.zi(), pos.getBlock(worldObj));

        TileEntity newTile = pos.getTileEntity(worldObj);

        if (newTile != null)
        {
            if (Loader.isModLoaded("BuildCraft|Factory"))
            {
                /** Special quarry compatibility code. */
                try
                {
                    Class clazz = Class.forName("buildcraft.factory.TileQuarry");

                    if (newTile.equals(clazz))
                    {
                        ReflectionHelper.setPrivateValue(clazz, newTile, true, "isAlive");
                    }
                }
                catch (Exception e)
                {
                    e.printStackTrace();
                }
            }
        }
    }

}

File: src/main/scala/edx/mechanical/mech/process/crusher/TileMechanicalPiston.scala
package edx.mechanical.mech.process.crusher

import java.lang.reflect.Method

import edx.core.Electrodynamics
import edx.mechanical.mech.TileMechanical
import net.minecraft.block.Block
import net.minecraft.block.material.Material
import net.minecraft.item.ItemStack
import net.minecraft.nbt.NBTTagCompound
import net.minecraft.tileentity.TileEntity
import net.minecraft.world.World
import resonant.api.recipe.{MachineRecipes, RecipeResource, RecipeType}
import resonant.lib.mod.config.Config
import resonant.lib.transform.vector.Vector3
import resonant.lib.utility.MovementUtility
import resonant.lib.utility.inventory.InventoryUtility

/**
 * Mechanical driven piston that can be used to move basic blocks and crush ores
 * @author Calclavia
 */
object TileMechanicalPiston
{
  @Config private var mechanicalPistonMultiplier: Int = 2
}

class TileMechanicalPiston extends TileMechanical(Material.piston)
{
  var markRevolve: Boolean = false

  //Constructor
  mechanicalNode = new NodeMechanicalPiston(this)
  isOpaqueCube = false
  normalRender = false
  customItemRender = true
  rotationMask = 63
  setTextureName("material_steel_dark")

  override def update
  {
    super.update
    if (markRevolve)
    {
      val movePosition: Vector3 = toVector3.add(getDirection)
      if (!hitOreBlock(movePosition))
      {
        if (!worldObj.isRemote)
        {
          val moveNewPosition: Vector3 = movePosition.clone.add(getDirection)
          if (canMove(movePosition, moveNewPosition))
          {
            move(movePosition, moveNewPosition)
          }
        }
      }
      markRevolve = false
    }
  }

  def hitOreBlock(blockPos: Vector3): Boolean =
  {
    val block: Block = blockPos.getBlock(world)
    if (block != null)
    {
      val blockStack: ItemStack = new ItemStack(block)
      val resources: Array[RecipeResource] = MachineRecipes.instance.getOutput(RecipeType.GRINDER.name, blockStack)
      if (resources.length > 0)
      {
        if (!worldObj.isRemote)
        {
          for (recipe <- resources)
          {
            if (Math.random <= recipe.getChance)
            {
              InventoryUtility.dropItemStack(world, blockPos.clone.add(0.5), recipe.getItemStack, 10, 0)
            }
          }
          blockPos.setBlockToAir(world)
        }
        Electrodynamics.proxy.renderBlockParticle(worldObj, blockPos.clone.add(0.5), new Vector3((Math.random - 0.5f) * 3, (Math.random - 0.5f) * 3, (Math.random - 0.5f) * 3), Block.getIdFromBlock(block), 1)
        return true
      }
    }
    if (!worldObj.isRemote)
    {
      world.destroyBlockInWorldPartially(0, blockPos.xi, blockPos.yi, blockPos.zi, -1)
    }
    return false
  }

  def canMove(from: Vector3, to: Vector3): Boolean =
  {
    if (this == to.getTileEntity(getWorldObj))
    {
      return false
    }
    val targetBlock: Block = to.getBlock(worldObj)
    if (!(worldObj.isAirBlock(to.xi, to.yi, to.zi) || (targetBlock != null && (targetBlock.canBeReplacedByLeaves(worldObj, to.xi, to.yi, to.zi)))))
    {
      return false
    }
    return true
  }

  def move(from: Vector3, to: Vector3)
  {
    val blockID: Block = from.getBlock(worldObj)
    val blockMetadata: Int = from.getBlockMetadata(worldObj)
    val tileEntity: TileEntity = from.getTileEntity(worldObj)
    val tileData: NBTTagCompound = new NBTTagCompound
    if (tileEntity != null)
    {
      tileEntity.writeToNBT(tileData)
    }
    MovementUtility.setBlockSneaky(worldObj, from, null, 0, null)
    if (tileEntity != null && tileData != null)
    {
      val isMultipart: Boolean = tileData.getString("id") == "savedMultipart"
      var newTile: TileEntity = null
      if (isMultipart)
      {
        try
        {
          val multipart: Class[_] = Class.forName("codechicken.multipart.MultipartHelper")
          val m: Method = multipart.getMethod("createTileFromNBT", classOf[World], classOf[NBTTagCompound])
          newTile = m.invoke(null, worldObj, tileData).asInstanceOf[TileEntity]
        }
        catch
          {
            case e: Exception =>
            {
              e.printStackTrace
            }
          }
      }
      else
      {
        newTile = TileEntity.createAndLoadEntity(tileData)
      }
      MovementUtility.setBlockSneaky(worldObj, to, blockID, blockMetadata, newTile)
      if (newTile != null && isMultipart)
      {
        try
        {
          val multipart: Class[_] = Class.forName("codechicken.multipart.MultipartHelper")
          multipart.getMethod("sendDescPacket", classOf[World], classOf[TileEntity]).invoke(null, worldObj, newTile)
          val tileMultipart: Class[_] = Class.forName("codechicken.multipart.TileMultipart")
          tileMultipart.getMethod("onMoved").invoke(newTile)
        }
        catch
          {
            case e: Exception =>
            {
              e.printStackTrace
            }
          }
      }
    }
    else
    {
      MovementUtility.setBlockSneaky(worldObj, to, blockID, blockMetadata, null)
    }
    notifyChanges(from)
    notifyChanges(to)
  }

  def notifyChanges(pos: Vector3)
  {
    worldObj.notifyBlocksOfNeighborChange(pos.xi, pos.yi, pos.zi, pos.getBlock(worldObj))
    /* val newTile: TileEntity = pos.getTileEntity(worldObj)
    if (newTile != null)
    {
        if (Loader.isModLoaded("BuildCraft|Factory"))
        {
            try
            {
                val clazz: Class[_ <: Any] = Class.forName("buildcraft.factory.TileQuarry")
                if (newTile == clazz)
                {
                    ReflectionHelper.setPrivateValue(clazz, newTile, true, "isAlive")
                }
            }
            catch
                {
                    case e: Exception =>
                    {
                        e.printStackTrace
                    }
                }
        }
    } */
  }
}